File: ch10/spring-data-example/src/main/java/com/packtpub/hands/on/reactive/kotlin/chapter10/springdataexample/CustomerController.java
package com.packtpub.hands.on.reactive.kotlin.chapter10.springdataexample;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.util.UriComponents;
import org.springframework.web.util.UriComponentsBuilder;

import java.util.Optional;
import java.util.stream.StreamSupport;

@RestController
public class CustomerController {
  private final CustomersRepository customersRepository;

  public CustomerController(CustomersRepository customersRepository) {
    this.customersRepository = customersRepository;
  }

  @GetMapping("/customer/{id}")
  ResponseEntity<?> getCustomer(@PathVariable Integer id) {
    final Optional<Customer> optionalCustomer = customersRepository.findById(id);
    if (optionalCustomer.isPresent()) {
      return ResponseEntity.ok(optionalCustomer.get());
    } else {
      return ResponseEntity.notFound().build();
    }
  }

  @PostMapping("/customer")
  ResponseEntity<?> postCustomer(@RequestBody Customer customer,
                                 UriComponentsBuilder builder) {
    final Customer savedCustomer = customersRepository.save(customer);
    final UriComponents uriComponents = builder.
        path("/customers/{id}").
        buildAndExpand(savedCustomer.getId());
    return ResponseEntity.created(uriComponents.toUri()).build();
  }

  @PatchMapping("/customer/{id}")
  ResponseEntity<?> patchCustomer(@PathVariable Integer id, @RequestBody Customer customer) {
    final Optional<Customer> possibleCustomer = customersRepository.findById(id);
    if (possibleCustomer.isPresent()) {
      final Customer foundCustomer = possibleCustomer.get();
      foundCustomer.setName(customer.getName());
      foundCustomer.setSurname(customer.getSurname());
      customersRepository.save(foundCustomer);
      return ResponseEntity.accepted().build();
    } else {
      return ResponseEntity.notFound().build();
    }
  }

  @DeleteMapping("/customer/{id}")
  ResponseEntity<?> deleteCustomer(@PathVariable Integer id) {
    final Optional<Customer> possibleCustomer = customersRepository.findById(id);
    if (possibleCustomer.isPresent()) {
      customersRepository.delete(possibleCustomer.get());
      return ResponseEntity.noContent().build();
    } else {
      return ResponseEntity.notFound().build();
    }
  }

  @GetMapping("/customers")
  ResponseEntity<?> getAllCustomers(){
    final Iterable<Customer> all = customersRepository.findAll();
    final long count = StreamSupport.stream(all.spliterator(), false).count();
    if(count>0){
      return ResponseEntity.ok(all);
    }else {
      return ResponseEntity.notFound().build();
    }
  }
}

File: ch11/spring-boot-kotlin-rest-api/src/main/kotlin/com/packtpub/hands/on/reactive/kotlin/chapter11/springbootkotlinrestapi/CustomerController.kt
package com.packtpub.hands.on.reactive.kotlin.chapter11.springbootkotlinrestapi

import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.util.UriComponentsBuilder

@RestController
class CustomerController(val customerRepository: CustomerRepository) {
  @GetMapping("/customers")
  fun getAllCustomers() = customerRepository.findAll().let {
    if (it.count() > 0) {
      ResponseEntity.ok(it)
    } else {
      ResponseEntity.notFound().build()
    }
  }

  @GetMapping("/customer/{id}")
  fun getCustomer(@PathVariable id: Int) = customerRepository.findById(id).let {
    if (it.isPresent) {
      ResponseEntity.ok(it.get())
    } else {
      ResponseEntity.notFound().build()
    }
  }

  @PostMapping("/customer")
  fun postCustomer(@RequestBody customer: Customer, builder: UriComponentsBuilder) =
      customerRepository.save(customer).let {
        ResponseEntity.created(builder.path("/customer/{id}").buildAndExpand(it._id).toUri()).build<Any>()
      }

  @DeleteMapping("/customer/{id}")
  fun deleteCustomer(@PathVariable id: Int) =
      if (customerRepository.findById(id).isPresent) {
        customerRepository.deleteById(id)
        ResponseEntity.noContent().build<Any>()
      } else {
        ResponseEntity.notFound().build<Any>()
      }
}

